device/pci_device.cpp

```diff
@@ -0,0 +1,431 @@
+    std::uint16_t pcie_domain;
+    std::uint8_t pcie_bus;
+    std::uint8_t pcie_device;
+    std::uint8_t pcie_function;
+
+    tenstorrent_get_device_info device_info;
+    memset(&device_info, 0, sizeof(device_info));
+    device_info.in.output_size_bytes = sizeof(device_info.out);
+    if (ioctl(this->device_fd, TENSTORRENT_IOCTL_GET_DEVICE_INFO, &device_info) == -1) {
+        TT_THROW("Get PCIe device info failed on device: ", this->device_id);
+    }
+    pcie_domain = device_info.out.pci_domain;
+    pcie_bus = device_info.out.bus_dev_fn >> 8;
+    pcie_device = PCI_SLOT(device_info.out.bus_dev_fn);
+    pcie_function = PCI_FUNC(device_info.out.bus_dev_fn);
+
+    // Get the PCIe revision ID from sysfs
+    static const char sys_pattern[] = "/sys/bus/pci/devices/%04x:%02x:%02x.%u/%s";
+    char buf[sizeof(sys_pattern) + 10];
+
+    // revision pattern = "/sys/bus/pci/devices/%04x:%02x:%02x.%u/revision"
+    std::snprintf(buf, sizeof(buf), sys_pattern, pcie_domain, pcie_bus, pcie_device, pcie_function, "revision");
+
+    std::ifstream revision_file(buf);
+    std::string revision_string;
+    if (std::getline(revision_file, revision_string)) {
+        this->pcie_device_id = device_info.out.device_id;
+        this->pcie_revision_id = std::stoi(revision_string, nullptr, 0);
+    } else {
+        TT_THROW("Revision ID /sys/ read failed for device: ", this->device_id);
+    }
+
+    // Get NUMA node from sysfs
+    // numa node pattern = "/sys/bus/pci/devices/%04x:%02x:%02x.%u/numa_node"
+    std::snprintf(buf, sizeof(buf), sys_pattern, pcie_domain, pcie_bus, pcie_device, pcie_function, "numa_node");
+
+    std::ifstream num_node_file(buf);
+    std::string numa_node_string;
+    if (std::getline(num_node_file, numa_node_string)) {
+        this->numa_node = std::stoi(numa_node_string, nullptr, 0);
+    } else {
+        TT_THROW("Numa node /sys/ read failed for device: ", this->device_id);
+    }
+}
+
+// Open a unique device_id per host memory channel (workaround for ttkmd < 1.21 support for more than 1 pin per fd)
+void TTDevice::open_hugepage_per_host_mem_ch(uint32_t num_host_mem_channels) {
+    for (int ch = 0; ch < num_host_mem_channels; ch++) {
+        log_debug(LogSiliconDriver, "Opening device_fd_per_host_ch device index: {} ch: {} (num_host_mem_channels: {})", device_id, ch, num_host_mem_channels);
+        int device_fd_for_host_mem = find_device(device_id);
+        if (device_fd_for_host_mem == -1) {
+            throw std::runtime_error(std::string("Failed opening a host memory device handle for device ") + std::to_string(device_id));
+        }
+        device_fd_per_host_ch.push_back(device_fd_for_host_mem);
+    }
+}
+
+tt::ARCH TTDevice::get_arch() const {
+    return arch;
+}
+
+template<typename T>
+T* TTDevice::get_register_address(std::uint32_t register_offset) {
+    void *reg_mapping;
+    if (system_reg_mapping != nullptr && register_offset >= system_reg_start_offset) {
```

Comment:
I think we can get rid of `system_reg_mapping` and go through BAR0 for Wormhole.  You don't have to change this now, just pointing it out as a future opportunity for simplification.